src/resources/uart_esp32.cc

```diff
@@ -17,95 +17,584 @@
+  if (!try_set_iomux_pin(rts_pin, SOC_UART_RTS_PIN_IDX)) {
+    gpio_hal_iomux_func_sel(GPIO_PIN_MUX_REG[rts_pin], PIN_FUNC_GPIO);
+    gpio_set_direction(rts_pin, GPIO_MODE_OUTPUT);
+    esp_rom_gpio_connect_out_signal(rts_pin, UART_PERIPH_SIGNAL(port_, SOC_UART_RTS_PIN_IDX), false, false);
+  }
+}
+
+void UartResource::set_cts_pin(gpio_num_t cts_pin) {
+  if (!try_set_iomux_pin(cts_pin, SOC_UART_CTS_PIN_IDX)) {
+    gpio_hal_iomux_func_sel(GPIO_PIN_MUX_REG[cts_pin], PIN_FUNC_GPIO);
+    gpio_set_pull_mode(cts_pin, GPIO_PULLUP_ONLY);
+    gpio_set_direction(cts_pin, GPIO_MODE_INPUT);
+    esp_rom_gpio_connect_in_signal(cts_pin, UART_PERIPH_SIGNAL(port_, SOC_UART_CTS_PIN_IDX), false);
+  }
+}
+
+UART_ISR_INLINE void UartResource::enable_interrupt_index_isr(uart_toit_interrupt_index_t index) {
+  IsrSpinLocker locker(&spinlock_);
+  enable_interrupt_mask_(interrupt_mask(index));
+}
+
+void UartResource::enable_interrupt_index(uart_toit_interrupt_index_t index) {
+  SpinLocker locker(&spinlock_);
+  enable_interrupt_mask_(interrupt_mask(index));
+}
+
+UART_ISR_INLINE void UartResource::disable_interrupt_index_isr(uart_toit_interrupt_index_t index) {
+  IsrSpinLocker locker(&spinlock_);
+  disable_interrupt_mask_(interrupt_mask(index));
+}
+
+UART_ISR_INLINE void UartResource::disable_read_interrupts() {
+  disable_interrupt_index_isr(UART_TOIT_INTR_RXFIFO_FULL);
+  disable_interrupt_index_isr(UART_TOIT_INTR_RX_TIMEOUT);
+}
+
+void UartResource::enable_read_interrupts() {
+  enable_interrupt_index(UART_TOIT_INTR_RXFIFO_FULL);
+  enable_interrupt_index(UART_TOIT_INTR_RX_TIMEOUT);
+}
+
+void UartResource::disable_interrupt_index(uart_toit_interrupt_index_t index) {
+  SpinLocker locker(&spinlock_);
+  disable_interrupt_mask_(interrupt_mask(index));
+}
+
+void UartResource::clear_interrupt_index(uart_toit_interrupt_index_t index) {
+  uart_toit_hal_clr_intsts_mask(hal_, interrupt_mask(index));
+}
+
+void UartResource::clear_interrupt_mask(uint32_t mask) {
+  uart_toit_hal_clr_intsts_mask(hal_, mask);
+}
+
+bool UartResource::try_set_iomux_pin(gpio_num_t pin, uint32_t iomux_index) const {
+  const uart_periph_sig_t *uart_pin = &uart_periph_signal[port_].pins[iomux_index];
+  if (uart_pin->default_gpio == -1 || uart_pin->default_gpio != pin) return false;
+  gpio_iomux_out(pin, uart_pin->iomux_func, false);
+  if (uart_pin->input) gpio_iomux_in(pin, uart_pin->signal);
+  return true;
+}
+
+size_t UART_ISR_INLINE UartResource::get_tx_fifo_available_count() {
+  return uart_toit_hal_get_txfifo_len(hal_);
+}
+
+UART_ISR_INLINE uart_event_types_t UartResource::interrupt_handler_read() {
+  size_t read_length = uart_toit_hal_get_rxfifo_len(hal_);
+  if (read_length == 0) return UART_EVENT_MAX;
+
+  size_t rx_buffer_free = rx_buffer_.free_count();
+  if (rx_buffer_free == 0) {
+    disable_read_interrupts();
+    return UART_BUFFER_FULL;
+  } else {
+    if (rx_buffer_free < read_length) read_length = rx_buffer_free;
+    uint8_t buffer[read_length];
```

Comment:
This is an alloca in an interrupt handler, which makes me very nervous.  What are the bounds on the size of this, and how big is an interrupt handler stack on ESP-IDF?  There is a new call in the pipeline that can help us avoid this issue: https://github.com/espressif/esp-idf/issues/10527

In the mean time perhaps a temp buffer in the resource is best?  Or perhaps a fixed size on-stack buffer with a loop that does uart_toit_hal_read_rxfifo and rx_buffer().send (aka xRingbufferSendFromISR) until it is finished?

Reply:
The max size is 128 bytes. It has been tested at very high speeds (3 Mbit), so I am not worried at all.

Reply:
Is it not limited by the max ring buffer size, 2048?

Reply:
Ah, no I see it's limited by the HAL. Perhaps insert an assert here that the size is no more than 256?